voting/orbs-pos-data/src/testkit/OrbsTokenServiceMock.ts

```diff
@@ -0,0 +1,187 @@
+import { PromiEvent, TransactionReceipt } from 'web3-core';
+import Web3PromiEvent from 'web3-core-promievent';
+import { IOrbsTokenService } from '../interfaces/IOrbsTokenService';
+
+export type OrbsAllowanceChangeCallback = (error: Error, allowance: string) => void;
+
+interface ITxData {
+  promiEvent: PromiEvent<TransactionReceipt> & {
+    resolve: (txReceipt: TransactionReceipt) => void;
+    reject: (reason: string) => void;
+  };
+  txReceipt: TransactionReceipt;
+}
+
+export class OrbsTokenServiceMock implements IOrbsTokenService {
+  private txDataMap: Map<object, ITxData> = new Map();
+  private addressToAllowancesMap: Map<string, Map<string, string>> = new Map();
+  private allowanceChangeEventsMap: Map<string, Map<string, Map<number, OrbsAllowanceChangeCallback>>> = new Map<
+    string,
+    Map<string, Map<number, OrbsAllowanceChangeCallback>>
+  >();
+
+  private senderAccountAddress: string = 'DUMMY_FROM_ADDRESS';
+
+  constructor(public autoCompleteTxes: boolean = true) {}
+
+  // CONFIG //
+  setFromAccount(address: string): IOrbsTokenService {
+    this.senderAccountAddress = address;
+    return this;
+  }
+
+  // WRITE //
+  approve(spenderAddress: string, amount: number): PromiEvent<TransactionReceipt> {
+    const promitEvent = this.generateTxData();
+    if (this.autoCompleteTxes) {
+      this.completeTx(promitEvent);
+      // TODO : O.L : Check if this is the best place to put this effect
```

Comment:
Implement  your idea regarding  map of effects

Reply:
Done